Source/MQTTnet/Internal/AsyncQueue.cs

```diff
@@ -24,22 +28,37 @@ public async Task<AsyncQueueDequeueResult<TItem>> TryDequeueAsync(CancellationTo
             {
                 try
                 {
-                    if (_semaphore == null)
+                    Task task;
+                    lock (_syncRoot)
```

Comment:
again why lock the whole purpose of _semaphore is to avoid lock. if you want to block just use lock and remove _semaphore

Reply:
The initial idea of the semaphore was to have a lock which supports async/await.  This is not possible with a classic lock. The thread will be waiting and not returned to the thread pool.